polar-core/src/terms.rs

```diff
@@ -0,0 +1,476 @@
+    pub args: TermList,
+}
+
+#[derive(Debug, Copy, Clone, Serialize, Deserialize, Eq, PartialEq, Hash)]
+pub enum Operator {
+    Debug,
+    Print,
+    Cut,
+    In,
+    Isa,
+    New,
+    Dot,
+    Not,
+    Mul,
+    Div,
+    Add,
+    Sub,
+    Eq,
+    Geq,
+    Leq,
+    Neq,
+    Gt,
+    Lt,
+    Unify,
+    Or,
+    And,
+    ForAll,
+    Assign,
+}
+
+impl Operator {
+    pub fn precedence(self) -> i32 {
+        match self {
+            Operator::Print => 11,
+            Operator::Debug => 11,
+            Operator::New => 10,
+            Operator::Cut => 10,
+            Operator::ForAll => 10,
+            Operator::Dot => 9,
+            Operator::In => 8,
+            Operator::Isa => 8,
+            Operator::Mul => 7,
+            Operator::Div => 7,
+            Operator::Add => 6,
+            Operator::Sub => 6,
+            Operator::Eq => 5,
+            Operator::Geq => 5,
+            Operator::Leq => 5,
+            Operator::Neq => 5,
+            Operator::Gt => 5,
+            Operator::Lt => 5,
+            Operator::Unify => 4,
+            Operator::Assign => 4,
+            Operator::Not => 3,
+            Operator::Or => 2,
+            Operator::And => 1,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq, Hash)]
+pub struct Operation {
+    pub operator: Operator,
+    pub args: TermList,
+}
+
+/// Represents a pattern in a specializer or after isa.
+#[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq, Hash)]
+pub enum Pattern {
+    Dictionary(Dictionary),
+    Instance(InstanceLiteral),
+}
+
+impl Pattern {
+    pub fn value_as_pattern(value: &Value) -> Value {
+        match value.clone() {
+            Value::InstanceLiteral(lit) => Value::Pattern(lit.as_pattern()),
+            Value::Dictionary(dict) => Value::Pattern(dict.as_pattern()),
+            v => v,
+        }
+    }
+
+    pub fn term_as_pattern(term: &Term) -> Term {
+        term.clone_with_value(Self::value_as_pattern(term.value()))
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
+pub enum Value {
+    Number(Numeric),
+    String(String),
+    Boolean(bool),
+    ExternalInstance(ExternalInstance),
+    // TODO (dhatch) Remove this type so that it is no longer possible to even make an
+    // instance literal value!
+    InstanceLiteral(InstanceLiteral),
+    Dictionary(Dictionary),
+    Pattern(Pattern),
+    Call(Predicate), // @TODO: Do we just want a type for this instead?
+    List(TermList),
+    Variable(Symbol),
+    RestVariable(Symbol),
+    Expression(Operation),
+}
+
+impl Value {
+    pub fn symbol(self) -> Result<Symbol, error::RuntimeError> {
+        match self {
+            Value::Variable(name) => Ok(name),
+            Value::RestVariable(name) => Ok(name),
+            _ => Err(error::RuntimeError::TypeError {
+                msg: format!("Expected symbol, got: {}", self.to_polar()),
+                stack_trace: None, // @TODO
```

Comment:
What is this TODO for? (and all of the identical ones below this in the `impl Value` block)

Reply:
I think it's the "Ensure all errors contain stack traces" task, right?

Reply:
If you think it would be worth the reminder, could you update the comments to something like
```suggestion
                stack_trace: None, // @TODO: Ensure all errors contain stack traces.
```